autopush/jwt.py

```diff
@@ -0,0 +1,102 @@
+    return string
+
+
+class VerifyJWT:
+
+    def __init__(self):  # pragma: no cover
+        pass
+
+    @classmethod
+    def raw_sig_to_der(cls, auth):
+        # type: (cls, str) -> str
+        """Fix the JWT auth token.
+
+        The `ecdsa` library signs using a raw, 32octet pair of values (r,s).
+        Cryptography, which uses OpenSSL, uses a DER sequence of (s, r).
+        This function converts the raw ecdsa to DER.
+
+        """
+        payload, asig = auth.rsplit(".", 1)
+        sig = base64.urlsafe_b64decode(repad(asig).encode('utf8'))
+        if len(sig) != 64:
+            return auth
+
+        # ecdsa and openssl transpose the "r" and "s" values of the signatures.
+        # for ecdsa signature is (r, s)
+        # for openssl, signature is (s, r)
+        # It's ok, though, because neither label them, even though openssl
+        # uses namedtypes, with the names set to ""
+        #
+        # Oh frabjous day!
+        ds = univ.Sequence(
+            componentType=namedtype.NamedTypes()
+        ).setComponents(
+            univ.Integer(int(binascii.hexlify(sig[:32]), 16)),
+            univ.Integer(int(binascii.hexlify(sig[32:]), 16)),
+        )
+        encoded = encode(ds)
+        new_sig = base64.urlsafe_b64encode(encoded)
+        return "{}.{}".format(payload, new_sig)
+
+    @classmethod
+    def decode(cls, token, key=None, *args, **kwargs):
+        # type (cls, str, str) -> dict()
+        """Decode a web token into a assertion dictionary.
+
+        This attempts to rectify both ecdsa and openssl generated
+        signatures. We use the built-in cryptography library since it wraps
+        libssl and is faster than the python only approach.
+
+        This raises an InvalidSignature exception if the signature fails.
+
+        """
+        # convert the signature if needed.
+        token = cls.raw_sig_to_der(token)
+        sig_material, signature = token.rsplit('.', 1)
+        pkey = ec.EllipticCurvePublicNumbers.from_encoded_point(
+            ec.SECP256R1(),
+            key
+        ).public_key(default_backend())
+        # eventually this would be a map of algs to signing algorithms.
+        # NOTE: verifier() will take any string as the signature. It appears
+        # to be doing lazy verification and matching strings rather than
+        # comparing content values. If the signatures start failing for
+        # some unknown reason in the future, decode the signature and
+        # make sure it matches how we're reconstructing it.
+        verifier = pkey.verifier(
+            base64.urlsafe_b64decode(repad(signature.encode())),
+            ec.ECDSA(hashes.SHA256()))
+        verifier.update(sig_material.encode())
+        # This will raise an InvalidSignature exception if failure.
+        # It will be captured externally.
+        verifier.verify()
+        return json.loads(
+            base64.urlsafe_b64decode(repad(sig_material.split('.')[1])))
```

Comment:
You can probably factor out the `base64.urlsafe_b64decode(repad(signature.encode()))` above into a `message` var, to safe teh re-decoding.